synapse/rest/media/v1/filepath.py

```diff
@@ -43,47 +43,79 @@ def _wrapped(self: "MediaFilePaths", *args: Any, **kwargs: Any) -> str:
+        relative: A boolean indicating whether the wrapped method returns paths relative
+            to the media store directory.
 
     Returns:
-        The method, wrapped with a check to ensure that the returned path(s) lie within
-        the media store directory. Raises a `ValueError` if the check fails.
+        A method which will wrap a path-returning method, adding a check to ensure that
+        the returned path(s) lie within the media store directory. The check will raise
+        a `ValueError` if it fails.
     """
 
-    @functools.wraps(func)
-    def _wrapped(
-        self: "MediaFilePaths", *args: Any, **kwargs: Any
-    ) -> Union[str, List[str]]:
-        path_or_paths = func(self, *args, **kwargs)
-
-        if isinstance(path_or_paths, list):
-            paths_to_check = path_or_paths
-        else:
-            paths_to_check = [path_or_paths]
-
-        for path in paths_to_check:
-            # path may be an absolute or relative path, depending on the method being
-            # wrapped. When "appending" an absolute path, `os.path.join` discards the
-            # previous path, which is desired here.
-            normalized_path = os.path.normpath(os.path.join(self.real_base_path, path))
-            if (
-                os.path.commonpath([normalized_path, self.real_base_path])
-                != self.real_base_path
-            ):
-                raise ValueError(f"Invalid media store path: {path!r}")
-
-        return path_or_paths
-
-    return cast(GetPathMethod, _wrapped)
+    def _wrap_with_jail_check_inner(func: GetPathMethod) -> GetPathMethod:
+        @functools.wraps(func)
+        def _wrapped(
+            self: "MediaFilePaths", *args: Any, **kwargs: Any
+        ) -> Union[str, List[str]]:
+            path_or_paths = func(self, *args, **kwargs)
+
+            if isinstance(path_or_paths, list):
+                paths_to_check = path_or_paths
+            else:
+                paths_to_check = [path_or_paths]
+
+            for path in paths_to_check:
+                # Construct the path that will ultimately be used.
+                # We cannot guess whether `path` is relative to the media store
+                # directory, since the media store directory may itself be a relative
+                # path.
+                if relative:
+                    normalized_path = os.path.normpath(
+                        os.path.join(self.base_path, path)
+                    )
+                else:
+                    normalized_path = os.path.normpath(path)
+
+                normalized_base_path = os.path.normpath(self.base_path)
+
+                # Check that the path lies within the media store directory.
+                # `os.path.commonpath` does not take `../`s into account and
+                # considers `a/b/c` and `a/b/c/../d` to have a common path of
+                # `a/b/c`, so we have to normalize the paths first.
```

Comment:
This comment makes it sound like we an't use `commonpath`, but then we *do* use it below... do we need to do something about these issues?

Reply:
It sounds like the comment could do with rewording. It was intended to explain why we use `normpath` before `commonpath`. 

As for the `normpath` issue below, I'm really struggling to find a way to resolve it.

As an example, if you have a media store at `/var/lib/synapse`, symlinked to `/mnt/media`, then `/var/lib/synapse/../synapse/some_file` refers to `/mnt/synapse/some_file` and not `/mnt/media/some_file`.
`normpath` will elide the `..` and make the path look like it's legitimate.

`realpath` would resolve the path correctly, but then server admins who have symlinks _within_ the media store directory would have problems, eg. if `remote_content` was symlinked elsewhere.

Since this check is a last resort and not _the_ check for path traversal, I opted to be permissive and let the issue slide.

(todo tomorrow: revisit comments)

Reply:
I wonder if it is reasonable to not allow relative paths or not allow symlinks somewhere in here?

Reply:
I agree, it'd be nice to not allow symlinks within the media store. On the other hand, I can definitely imagine a server admin wanting to offload one of the big media subdirectories elsewhere and expecting a symlink to work.

As for relative paths (paths containing `..` or `.`), in theory we _shouldn't_ encounter any if the main checks work. Then again this decorator is wholly unnecessary if the main checks work.

Reply:
I've tried to reword the comments. Hopefully they are less confusing